Learning emails 4534

rae-stanton opened the pull request:
### What GitHub issue is this PR for, if any?
Resolves #4534 

### What changed, and why?
Added new learning hour report functionality for supervisors and admins to receive on a monthly basis. This is available for selection for admin edit and supervisor edit pages. 

### How will this affect user permissions?
- Volunteer permissions: N/A
- Supervisor permissions: Adds toggle for monthly learning hour report email for their org
- Admin permissions: Adds toggle for monthly learning hour report email for their org - can also add to their respective supervisors on their edit pages.

### How is this tested? (please write tests!) 💖💪
Mailer spec test added for core mailer, happy to add extra test coverage, but I wasn't sure if tests on the small model changes for admins/supervisors were needed to reflect toggling receiving learning hours reports monthly. I don't have access to Heroku for seeing if the cron jobs are working as expected, so that may need to be tested out by someone with Heroku access. 

### Screenshots please :)
<img width="1107" alt="Screenshot 2023-09-05 at 9 25 45 AM" src="https://github.com/rubyforgood/casa/assets/101673900/eb17941a-a4c3-42cb-ab19-81c655a83bfa">
<img width="1432" alt="Screenshot 2023-09-05 at 9 30 08 AM" src="https://github.com/rubyforgood/casa/assets/101673900/2517b724-7e68-4a65-8fad-e3979237102f">

spaulsandhu committed:
WIP

spaulsandhu committed:
WIP

rae-stanton committed:
Merge pull request #1 from spaulsandhu/learning-emails-4534

WIP

rae-stanton committed:
remove pry

rae-stanton committed:
remove pry

rae-stanton committed:
schema incoming

rae-stanton committed:
schema incoming

rae-stanton committed:
update migration

rae-stanton committed:
Merge branch 'learning-emails-4534' of https://github.com/rae-stanton/casa into learning-emails-4534

rae-stanton committed:
Update supervisors_controller.rb

rae-stanton committed:
Initial mailer spec

rae-stanton committed:
lint

rae-stanton commented:
CodeClimate seems to be complaining about duplicated code in some mailers - I guess that's unavoidable here.

littleforest commented on app/views/learning_hours_mailer/learning_hours_report_email.html.erb:
```diff
@@ -0,0 +1,80 @@
+                      <tr style="box-sizing: border-box;">
+                        <td class="card-content" style="box-sizing: border-box; font-size: 13px; line-height: 20px; color: rgb(111, 119, 125); padding-top: 10px; padding-right: 20px; padding-bottom: 0px; padding-left: 20px; vertical-align: top;" valign="top">
+                          <h3 style="box-sizing: border-box; font-weight: 800; margin: 20px 0 5px; font-size: 16px;">Learning hours have been added this week!</h3>
+                          <p style="box-sizing: border-box;">Please see the attached full report or export the full report at <a href="https://casa-qa.herokuapp.com/reports" style="box-sizing: border-box;">https://casa-qa.herokuapp.com/reports</a>.</p>
```
We should use routes here so we don't end up with the QA site hardcoded -- `link_to reports_url reports_url`

rae-stanton replied:
got it!

littleforest commented on db/schema.rb:
```diff
@@ -583,6 +583,7 @@
     t.datetime "confirmation_sent_at"
     t.string "unconfirmed_email"
     t.string "old_emails", default: [], array: true
+    t.boolean "monthly_learning_hours_report", default: false
```
Can you also add `null: false` here?

littleforest commented on lib/tasks/monthly_learning_hours_report.rake:
```diff
@@ -0,0 +1,13 @@
+  supervisors = Supervisor.active.where(monthly_learning_hours_report: true)
+
+  admins.each do |admin|
+    LearningHoursMailer.learning_hours_report_email(admin).deliver
```
```suggestion
    LearningHoursMailer.learning_hours_report_email(admin).deliver_later
```

littleforest commented on lib/tasks/monthly_learning_hours_report.rake:
```diff
@@ -0,0 +1,13 @@
+  end
+
+  supervisors.each do |supervisor|
+    LearningHoursMailer.learning_hours_report_email(supervisor).deliver
```
```suggestion
    LearningHoursMailer.learning_hours_report_email(supervisor).deliver_later
```

littleforest commented on app/mailers/learning_hours_mailer.rb:
```diff
@@ -0,0 +1,24 @@
+  def learning_hours_report_email(user)
+    # user is either an Admin or Supervisor, this mailer is invoked through the rake task :monthly_learning_hours_report.rake
+
+    @user = user
```
Are you using `@user` in the mailer at all? If not, this doesn't need to be an instance variable, and you can just use `user` everywhere in this method.

rae-stanton replied:
Yeah I am using `@user` in a couple of places throughout the mailer. (:

littleforest commented on app/mailers/learning_hours_mailer.rb:
```diff
@@ -0,0 +1,24 @@
+    if @user.nil?
+      Bugsnag.notify("No user provided for LearningHoursMailer#learning_hours_report_email")
+      return
+    end
```
I don't feel like this is really necessary. It doesn't seem like anyone would write code that would not pass a user, and if they did, we could just let it fail at line 12 and figure out what is going on from the backtrace. This is getting called in a background job, so it's fine to let it fail without catching the error.

rae-stanton replied:
Gotcha, I went ahead and removed this from the mailer & the test.

littleforest commented on app/views/learning_hours_mailer/learning_hours_report_email.html.erb:
```diff
@@ -0,0 +1,80 @@
+                      <tr style="box-sizing: border-box;">
+                        <td class="cell c1776" style="box-sizing: border-box; width: 70%; vertical-align: middle;" width="70%" valign="middle">
+                          <div class="c1144" style="box-sizing: border-box; padding-top: 10px; padding-right: 10px; padding-bottom: 10px; padding-left: 10px; font-size: 17px; font-weight: 500;text-align: center;color: #fff;">Court
+                            Appointed Special Advocate (CASA) / <%= @casa_organization&.name %>
```
It looks like you've defined `@casa_org` in the mailer, but you are using `@casa_organization` here. Since the organization should always exist, you don't need `&.` and can just do `@casa_organization.name`.

rae-stanton replied:
Yeah I used `@casa_org` so I changed that to match in the view here so it renders correctly.

rae-stanton replied:
I went ahead and changed this in both spots to reflect what's in my mailer. (:

littleforest commented on app/views/learning_hours_mailer/learning_hours_report_email.html.erb:
```diff
@@ -0,0 +1,80 @@
+                  <div class="c2577" style="text-align:center;box-sizing: border-box; padding-top: 10px; padding-right: 10px; padding-bottom: 10px; padding-left: 10px;">
+                    <p class="footer-info" style="box-sizing: border-box;">Court Appointed Special Advocate (CASA) / <%= @casa_organization&.name %>
+                      <br style="box-sizing: border-box;">
+                      <br style="box-sizing: border-box;"><%= @casa_organization&.address %>
```
Fix `@casa_org` vs `@casa_organization` here too. Don't need `&.` here either.

littleforest reviewed (commented):
Thanks so much for working on this! Looking good.

rae-stanton committed:
Adjustments for mailer view

Removed some extra cruft from the mailer view, updated mailer, and updated rake task

rae-stanton committed:
adds null:false to db

rae-stanton committed:
removed bugsnag from test

rae-stanton commented:
@littleforest I updated the commits here - thanks for reviewing it. Let me know if anything else needs adjusting.

littleforest reviewed (approved):
🎉

littleforest commented:
@rae-stanton can you merge in main and push back out? Let's see if those flaky tests will pass.

rae-stanton commented:
> @rae-stanton can you merge in main and push back out? Let's see if those flaky tests will pass.

Yeah! Just making sure - you want me to go ahead and run the suite while in main and make sure they're passing?

rae-stanton committed:
Merge branch 'main' into learning-emails-4534

littleforest commented:
@rae-stanton no, I meant can you merge the `main` branch into your branch (or upstream `main` since it looks like you are working on a fork), and push it back out to GitHub?

rae-stanton committed:
Merge remote-tracking branch 'upstream/main' into learning-emails-4534

rae-stanton commented:
@littleforest I gotcha! I went ahead and upstreamed `main` into my fork. Would you like me to make any other test adjustments or anything else? Thanks for reviewing this, by the way! 😊
